### DISCUSSION TITLE
Auto increment asset tag?

### DISCUSSION THREAD

**Discussion by sliddjur** [netbox-community/netbox#9226] [Q&A] [answered]
https://github.com/netbox-community/netbox/issues/7501

I saw this was discussed in this issue, and I am wondering now with version 3.2.1, is this easily achievable without a plugin, or if someone already has developed a plugin? @PieterL75 ? :)

**Comment by kkthxbye-code** [comment, marked as answer]
You could do it with custom validators which would save you creating a plugin. If you provide your exact usecase, I can give you an example if you are not comfortable in python.

**Reply by sliddjur** [reply]
The issue of incrementing the asset tag manually is that its not very smooth. One has to go and list all devices, show the asset-tag column, and sort by asset tag, them manually select the next asset tag.

The use case would be to chose a starting asset-tag number, and then when the next device is created, just auto-increment it by 1.
We name all our devices by our own asset names and tags with "datacenter-####".

eg: 
datacenter-5001
datacenter-5002
datacenter-5003

Device42 has an even more complex way of doing this with templates. This is not needed in the first place.
https://docs.device42.com/mobile-inventory/auto-assign-asset-numbers/

**Reply by kkthxbye-code** [reply]
```
from extras.validators import CustomValidator

class MyValidator(CustomValidator):

    def validate(self, instance):
        from dcim.models import Device
        prefix = "datacenter-"
        max_digits = 5

        if not instance.asset_tag:
            device = Device.objects.filter(asset_tag__startswith=prefix).order_by("-name").first()

            if not device:
                # Return 1 if no device has an asset tag yet.
                instance.asset_tag = prefix + "1".zfill(max_digits)
                return

            try:
                next_id = int(device.asset_tag.removeprefix(prefix)) + 1
                instance.asset_tag = prefix + str(next_id).zfill(max_digits)
            except ValueError:
                # If we fail to parse the id do nothing (proper error handling can be added)
                return

CUSTOM_VALIDATORS = {
    'dcim.device': (
        MyValidator(),
    )
}
```

Here's a working example. I wasn't sure if the datacenter- prefix is static or you have multiple datacenters with different prefixes, but the code should serve as an example.

First time you create or update a device with no asset_tag it will get the tag datacenter-00001. Next device you save it will look up all devices with an asset tag starting with datacenter-, then parse the number, add 1 and modify the tag before saving. As a result the second device will get datacenter-00002.

If you want to test it you can paste it in the bottom of your configuration.py file.

More documentation here:
https://demo.netbox.dev/static/docs/customization/custom-validation/

**Reply by sliddjur** [reply]
Thanks!
I created one device without asset-tag, and I got the asset-tag datacenter-00001
The second device without asset-tag I created gave me this error

```
Server Error
There was a problem with your request. Please contact an administrator.

The complete exception is provided below:

<class 'AttributeError'>

'str' object has no attribute 'removeprefix'

Python version: 3.8.10
NetBox version: 3.2.1
If further assistance is required, please post to the [NetBox discussion forum](https://github.com/netbox-community/netbox/discussions) on GitHub.
```

**Reply by kkthxbye-code** [reply]
I'm on python 3.9 so I used the removeprefix method. Here's a version that should work on python 3.8:

```
from extras.validators import CustomValidator

class MyValidator(CustomValidator):

    def validate(self, instance):
        from dcim.models import Device
        prefix = "datacenter-"
        max_digits = 5

        if not instance.asset_tag:
            device = Device.objects.filter(asset_tag__startswith=prefix).order_by("-asset_tag").first()

            if not device:
                # Return 1 if no device has an asset tag yet.
                instance.asset_tag = prefix + "1".zfill(max_digits)
                return

            try:
                next_id = int(device.asset_tag[len(prefix):]) + 1
                instance.asset_tag = prefix + str(next_id).zfill(max_digits)
            except ValueError:
                # If we fail to parse the id do nothing (proper error handling can be added)
                return

CUSTOM_VALIDATORS = {
    'dcim.device': (
        MyValidator(),
    )
}
```

**Reply by jeremystretch** [reply]
This is a really interesting use for custom validators that I hadn't considered: Using them to automatically set undefined instance attributes. Might be worth calling out as an example in the docs.

**Reply by sliddjur** [reply]
Thanks, now the second device works.

But one issue is that when I edit an already existing device without an asset tag, and press save, I get this error.

```
Errors
Asset tag
Device with this Asset tag already exists.
```

Also, it kind of breaks my scripts that runs to update and create some devices. Is it perhaps possible to make a checkbox in the GUI "generate asset tag" on the "create device" screen that will call this script?

Again, thanks for your time.

**Reply by kkthxbye-code** [reply]
> But one issue is that when I edit an already existing device without an asset tag, and press save, I get this error.

Sorry, that was a typo from my end, sorting by device name instead of asset_tag. I have edited the second version.

> Also, it kind of breaks my scripts that runs to update and create some devices.

Should also be fixed by my new version I guess.

@jeremystretch - Yeah, it's like a light version of post_save signals.

**Comment by LHBL2003** [comment]
We have inventory numbers starting from SP0001 to SP4300 something. Of course I don't want any problems when it gets to 5 digits. Therefore, I have adapted it a little so that it also works with a number caos.

Example alphabetical sorting:
- SP4348
- SP4338
- SP40002
- SP40001
- SP40000
- SP1000
- SP0303
- SP03001
- SP0001

Next Result: SP40003

```
###################################
# DE: Ist kein Asset-Tag angegeben, wird das höchste Asset-Tag beginnend mit SP gesucht und um 1 erhöht. Gibt es noch kein Asset Tag so wird mit 1 gestartet.
# EN: If no asset tag is specified, the highest asset tag starting with SP is searched for and incremented by 1. If there is no asset tag yet, it is started with 1.

# THX: https://github.com/netbox-community/netbox/discussions/9226
#      https://stackoverflow.com/a/70153906/27018495
class CreateAssetTag(CustomValidator):

    def validate(self, obj):        
        from django.db.models.expressions import RawSQL
        from operator import attrgetter
        
        prefix = "SP" # Text that should appear in the asset name on the left
        max_digits = 0 # Number of leading zeros in the count
       
        if not obj.asset_tag:
            compareObj = obj.__class__.objects.filter(asset_tag__iregex=r'^' + prefix + r'(\d+)$')\
                        .all().annotate(theLastAssetId=RawSQL("CAST(regexp_replace(asset_tag, '[^0-9]+', '', 'g') AS INTEGER)", '')) #.order_by('-theLastAssetId').first() # *Alternative to the max function
               
            maxObj = max(compareObj, key=attrgetter('theLastAssetId')) # Entry with the highest asset value
               
            if not compareObj:
                # Returns 1 if no comparison object has an asset tag yet.
                obj.asset_tag = prefix + "1".zfill(max_digits)
                return

            try:
                next_id = maxObj.theLastAssetId + 1
                #next_id = compareObj.theLastAssetId + 1 # *Alternative to the max function
                obj.asset_tag = prefix + str(next_id).zfill(max_digits)
                
            except ValueError:
                # If the ID cannot be parsed, nothing happens (appropriate error handling can be added)
                return
```

It can certainly be optimized. :)
But I'm glad that it worked out on Friday. 
~~I wanted to see if you can replace Order by with Max. It's certainly quicker than sorting.~~
Addendum: order_by has been replaced by the max function. But left as a comment as a teaching factor.
